guides/source/tuning_perf_for_deployment.md

```diff
@@ -0,0 +1,209 @@
+
+The way to achieve true parallelism with Ruby is to use multiple processes. As long as there is a free CPU core, Ruby processes don't have to wait on one another before resuming execution after an I/O operation is complete.
+However, processes only share a fraction of their memory via [copy-on-write](https://en.wikipedia.org/wiki/Copy-on-write), so one additional process uses more memory than an additional thread would.
+
+However, note that while threads are cheaper than processes, they are not free, and increasing the number of threads per process, also increases memory usage.
+
+#### Practical Implications
+
+Users interested in optimizing for throughput and server utilization will want to run one process per CPU core and increase the number of threads per process until the impact on latency is dimmed too important.
+
+Users interested in optimizing for latency will want to keep the number of threads per process low.
+To optimize for latency even further, users can even set the thread per process count to `1` and run `1.5` or `1.3` process per CPU core to account for when processes are idle waiting for I/O operations.
+
+It is important to note that some hosting solutions may only offer a relatively small amount of memory (`RAM`) per CPU core, preventing you from running as many processes as needed to use all CPU cores,
+However, most hosting solutions have different plans with different ratios of memory and CPU. 
+
+Another thing to consider is that Ruby memory usage benefits from economies of scale thanks to [copy-on-write](https://en.wikipedia.org/wiki/Copy-on-write).
+So `2` servers with `32` Ruby processes each will use less memory per CPU core than `16` servers with `4` Ruby processes each.
+
+### Configurations
+
+#### Puma
+
+The Puma configuration resides in the `config/puma.rb` file.
+The two most important Puma configurations are the number of threads per process, and the number of processes, which Puma calls `workers`.
+
+The number of threads per process is configured via the `thread` directive.
+In the default generated configuration, it is set to `3`.
+You can modify it either by setting the `RAILS_MAX_THREADS` environment variable or simply editing the configuration file.
+
+The number of processes is configured by the `workers` directive.
+If you use more than one thread per process, then it should be set to how many CPU cores are available on the server, or if the server is running multiple applications, to how many cores you want the application to use.
+If you only use one thread per worker, then you can increase it to above one per process to account for when workers are idle waiting for I/O operations.
+In the default generated configuration, it is set to use all the available processor cores on the server via the `Concurrent.available_processor_count` helper. You can also modify it by setting the `WEB_CONCURRENCY` environment variable.
```

Comment:
Note: the part about `Concurrent.available_processor_count` isn't true yet.

The feature was added in https://github.com/ruby-concurrency/concurrent-ruby/pull/1038 but is yet unreleased. I reached the gem owners to try to get a release, if we could fix this it would remove the need to manually configure the processor count, which would be very nice.

cc @rafaelfranca